src/com/haskforce/actions/HaskellHindentFormatAction.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.haskforce.actions;
+
+/*
+ * Adapted from the Stylish Haskell formatter
+ */
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+
+import org.apache.sanselan.util.IOUtils;
+import org.jetbrains.annotations.NotNull;
+
+import com.haskforce.psi.HaskellFile;
+import com.haskforce.settings.ToolKey;
+import com.haskforce.utils.HaskellToolsNotificationListener;
+import com.intellij.execution.configurations.GeneralCommandLine;
+import com.intellij.execution.process.CapturingProcessAdapter;
+import com.intellij.execution.process.OSProcessHandler;
+import com.intellij.execution.process.ProcessEvent;
+import com.intellij.notification.Notification;
+import com.intellij.notification.NotificationType;
+import com.intellij.notification.Notifications;
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.actionSystem.CommonDataKeys;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.command.CommandProcessor;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.project.DumbAware;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiFile;
+import com.intellij.util.ExceptionUtil;
+
+/**
+ * Action that calls hindent on the buffer it is invoked for.
+ */
+public class HaskellHindentFormatAction extends AnAction implements DumbAware {
+    private static final String NOTIFICATION_TITLE = "Reformat code with Hindent";
+    private static final Logger LOG = Logger.getInstance(HaskellHindentFormatAction.class);
+
+    /**
+     * Enable the action for Haskell files.
+     */
+    @Override
+    public void update(AnActionEvent e) {
+        PsiFile psiFile = e.getData(CommonDataKeys.PSI_FILE);
+        boolean isHaskell = psiFile instanceof HaskellFile;
+        e.getPresentation().setEnabled(isHaskell);
+    }
+
+    /**
+     * Main entry point. Calls hindent and detects errors.
+     */
+    @Override
+    public void actionPerformed(AnActionEvent e) {
+        final PsiFile psiFile = e.getData(CommonDataKeys.PSI_FILE);
+        final Project project = getEventProject(e);
+        if (project == null) return;
+        if (!(psiFile instanceof HaskellFile)) return;
+        VirtualFile virtualFile = psiFile.getVirtualFile();
+        if (virtualFile == null) return;
+
+        final String groupId = e.getPresentation().getText();
```

Comment:
It seems `groupId` could be null here, but the Stylish action doesn't do anything with the possible null either, so maybe it's ok.  We could probably address this later and clean up both implementations.